c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+// `kj::_::Coroutine<Own<AsyncIoStream>>`.
+//
+// The implementation object lives in the heap-allocated coroutine frame. It gets destroyed and
+// deallocated when the frame does.
+
+namespace kj::_ {
+
+namespace stdcoro = KJ_COROUTINE_STD_NAMESPACE;
+
+template <typename Self, typename T>
+class CoroutineBase;
+// CRTP mixin, covered later.
+
+template <typename T>
+class Coroutine: public CoroutineBase<Coroutine<T>, T> {
+  // The standard calls this the `promise_type` object. We can call this the "coroutine
+  // implementation object" since the word promise means different things in KJ and std styles. This
+  // is where we implement how a `kj::Promise<T>` is returned from a coroutine, and how that promise
+  // is later fulfilled. We also fill in a few lifetime-related details.
+  //
+  // The implementation object is also where we can customize memory allocation of coroutine frames,
+  // by implementing a member `operator new(size_t, Args...)` (same `Args...` as in
+  // coroutine_traits).
+  //
+  // We can also customize how await-expressions are transformed within `kj::Promise<T>`-based
+  // coroutines by implementing an `await_transform(P)` member function, where `P` is some type for
+  // which we want to implement co_await support, e.g. `kj::Promise<U>`. This feature is useful
+  // to provide a more optimized `kj::EventLoop` integration when the coroutine's return type and
+  // the await-expression's type are both `kj::Promise` instantiations.
+
+public:
+  using Handle = stdcoro::coroutine_handle<Coroutine<T>>;
+
+  Coroutine() = default;
+  KJ_DISALLOW_COPY(Coroutine);
+
+  ~Coroutine() { if (adapter) adapter->coroutineDestroyed(); }
+  // The implementation object (*this) lives inside the coroutine's frame, so if it is being
+  // destroyed, then that means the frame is being destroyed. Since the frame might be destroyed
+  // *before* the adapter (e.g., the coroutine resumes from its last suspension point before the
+  // user destroys the owning promise), we need to tell the adapter not to try to double-free the
+  // frame with coroutine.destroy().
+
+  Promise<T> get_return_object() {
+    // Called after coroutine frame construction and before initial_suspend() to create the
+    // coroutine's return object.
+    return newAdaptedPromise<T, Adapter>(Handle::from_promise(*this));
+  }
+
+  auto initial_suspend() { return stdcoro::suspend_never(); }
+  auto final_suspend() noexcept { return stdcoro::suspend_never(); }
+  // These adjust the suspension behavior of coroutines immediately upon initiation, and immediately
+  // after completion.
+  //
+  // The initial suspension point could allow us to defer the initial execution of a coroutine with
+  // evalLater(), for instance.
+  //
+  // The final suspension point could be useful to delay deallocation of the coroutine frame to
+  // match the lifetime of the enclosing promise.
+
+  void unhandled_exception() {
```

Comment:
I guess we should wrap this in `#if !KJ_NO_EXCEPTIONS`? (Or otherwise wrap the entire coro implementation in it.)

Reply:
I tried to make coroutines work with -fno-exceptions, but got confused and tired and gave up: https://github.com/capnproto/capnproto/pull/1104/commits/23721723d988ea9a523e10e3543b6f5a3d7f0f18

I just gated the entire KJ_HAS_COROUTINE define behind !KJ_NO_EXCEPTIONS.